### ISSUE TITLE
Remove space between anonymous function keyword and parenthesis

### ISSUE THREAD

**Issue by tonix-tuft** [prettier/prettier#7972]
**Input:**

```jsx
  const callback = function(...args) {
    if (firstOperation) {
      firstOperation = false;
      onceCallback.apply(this, args);
    }
  };
```

**Output:**

```jsx
  const callback = function (...args) { // <--- Prettier adds a space between "function" and "("
    if (firstOperation) {
      firstOperation = false;
      onceCallback.apply(this, args);
    }
  };
```

**Expected behavior:**
Do not add the space, at least, it was not added before in previous versions.
Is there an option allowing me to keep the old behaviour?

Thanks!

**Comment by sosukesuzuki** [comment]
This is a specification discussed in #3847 and added via #3903  since 2.0. There is no option to disable this.

**Comment by tonix-tuft** [comment]
Thanks for your reply, but why am I experiencing this only now? Was this feature added last week?

**Comment by fisker** [comment]
Since `v2.0.0`

**Comment by tonix-tuft** [comment]
OK, thanks!

**Comment by harbulot** [comment]
@sosukesuzuki It would be good to have an option for this.

The [philosophy of having as few options as possible](https://prettier.io/docs/en/option-philosophy.html) comes from a good intention, but it has its limits.

The problem users now have is that, by just formatting existing files with the new version, they'll now have meaningless lines changes in their diffs/code repositories. For many of us, one of the main purposes of code formatting (at least harmonised across a project) is to make sure only the significant changes stand out when using revision control systems.

Suddenly having many diffs like this example across a project makes no sense:

```diff
-$(document).ready(function() {
+$(document).ready(function () {
```

Making changes that break the behaviour of past versions seems to go completely against the rationale for using a code formatter in the first place.

**Comment by tonix-tuft** [comment]
I agree with @harbulot, even if you change the default behaviour, leaving the option of configuring so that everything works as it used to would be a good idea.

**Comment by rwwagner90** [comment]
Would love to see an option to get back the old behavior. I don't personally like the space.

**Comment by tonix-tuft** [comment]
@rwwagner90 would love to see an option too

**Comment by evilebottnawi** [comment]
No options, now it is default behavior and already was discussed long time ago

**Comment by rwwagner90** [comment]
Why no options though? Many of us liked the way it worked in Prettier 1.x.

**Comment by evilebottnawi** [comment]
Now it is default behavior

**Comment by rwwagner90** [comment]
@evilebottnawi yes, I know, but it should be configurable, rather than a forced change.

**Comment by tonix-tuft** [comment]
I agree with @rwwagner90, but, yeah, Prettier is opinionated.

**Comment by harbulot** [comment]
@evilebottnawi The problem with that "opinionated" stuff is that it now goes against the goal of Prettier.

The [main goal of Prettier is to provide consistency](https://prettier.io/docs/en/index.html): "*It removes all original styling and ensures that all outputted code conforms to a **consistent** style.*" (and "*Prettier gets rid of all original styling and **guarantees consistency***" in the associated blog article).

How does it achieve that consistency? By being opinionated. Fair enough.

When I started to use Prettier, there were some choices I didn't necessarily like or would have made differently, but going along with the Prettier non-configurable choices was worth it, to achieve consistency.

Being opinionated was a worthwhile means to achieve the goal: consistency.

It looks like in the evolution towards version 2.0, and with the success of Prettier (which is deserved!), the means was allowed to overshadow the end goal.
By moving from Prettier 1.x to 2.0, that consistency is now lost. The code formatted now with Prettier is inconsistent (and cannot be made consistent) with what was there just a couple of months ago.

I can't speak for other users, but I don't really care either way about the space in this position. I have a mild preference for not having a space I suppose, but that could be by habit.
What I'm more concerned about is breaking the Git history simply for formatting. (I'm not sure how many people who suggest that all we have to do is to is something like `prettier --write "**/*.js"` have to deal with projects that have been developed over a period of time.)

**Comment by sarimarton** [comment]
I'd like to chime in in favor of the new rule. The new space rule actually _adds_ consistency in the rule set, as it fits in the more general tendency to add space around all keywords. Introducing it late in the game is not a problem, because it's a one-time change, you can easily run it through the code base and submit in a single commit, or fix the prettier version in the project (which is nonetheless a good practice).

**Comment by tonix-tuft** [comment]
FWIW, I'm unopinionated about Prettier's opinionated choices, so for me it really doesn't change much.

**Comment by harbulot** [comment]
@sarimarton I don't actually mind either way whether there is a space or not. You're saying it adds consistency, and that may indeed be the case, but that's what I'd call *cosmetic* consistency, at the expense of *historical* consistency.

Both are good, but my argument is that the former is less important than the latter. It may "just" be a one-time change that can be done in a single commit, but it messes up historical diffs (between two commits on before and after that commit) and Git blame (and in a lot of places, since anonymous functions are quite common in JS). As much as I like tidy and "pretty" code, messing up with these, which are very useful for bug investigations, is a bigger problem: this becomes functional, not just cosmetic.

Of course, as you suggest, we can fix the version of Prettier to a pre-2.0 version, but there's a chance they'll no longer work at some point, especially when integrated within plugins such as the one for VScode.

If the space is the new default, so be it. What I'm advocating is the introduction of an option to be able to keep the previous behaviour. While additional options don't quite go along with the opinionated philosophy of Prettier, it has the advantage of letting people stick with Prettier's previous choices if they want to. That would seem like a reasonable trade-off. If Prettier's asking its user to follow its opinions, it would also be good to have the ability to keep those opinions stable, even if the default approach changes later. Otherwise, it's almost as if we were switching to a different formatting tool altogether.
